src/ssl.c

```diff
@@ -48370,6 +48370,357 @@ int wolfSSL_PEM_write_bio_PKCS7(WOLFSSL_BIO* bio, PKCS7* p7)
     }
     return WOLFSSL_FAILURE;
 }
+
+/******************************************************************************
+* wolfSSL_SMIME_read_PKCS7 - Reads the given S/MIME message and parses it into a PKCS7 object.
+* In case of a multipart message, stores the signed data in bcont.
+*
+* RETURNS:
+* returns pointer to a PKCS7 structure on success, otherwise returns NULL
+*/
+WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in, WOLFSSL_BIO** bcont)
+{
+    MimeHeader* allHeaders = NULL;
+    MimeHeader* curHeader = NULL;
+    MimeParameter* curParameter = NULL;
+    int inLen = 0;
+    int sectionLen = 0;
+    int ret = -1;
+    char* section = NULL;
+    PKCS7* pkcs7 = NULL;
+
+
+    inLen = wolfSSL_BIO_get_len(in);
+    if (inLen <= 0) return NULL;
+
+    ret = MIME_parse_headers(in, inLen, &allHeaders);
+    if (ret < 0) {
+        WOLFSSL_MSG("Parsing MIME headers failed.\n");
+        return NULL;
+    }
+
+    curHeader = MIME_find_header_name("Content-Type", allHeaders);
+    if (curHeader && !XSTRNCMP(curHeader->body, "multipart/signed", XSTRLEN("multipart/signed"))) {
+        curParameter = MIME_find_param_attr("protocol", curHeader->parameters);
+        if (curParameter && (!strcmp(curParameter->value, "application/pkcs7-signature") ||
+                             !strcmp(curParameter->value, "application/x-pkcs7-signature"))) {
+            curParameter = MIME_find_param_attr("boundary", curHeader->parameters);
+            if (!curParameter) return NULL;
+
+            int lineLen = 0;
+            int remainLen = wolfSSL_BIO_get_len(in);
+            if (remainLen <= 0) return NULL;
+            section = XMALLOC((remainLen+1)*sizeof(byte), NULL, DYNAMIC_TYPE_PKCS7);
```

Comment:
Please cast all XMALLOC calls back to type as per coding standards.  Ex: (char*)XMALLOC(...).  Applies to multiple places in this PR.